Add missing keys to `CollatorPreferences`

jedel1043 opened the pull request:
Adds support for `kn` and `kf` on `CollatorPreferences`.

hsivonen commented:
Looks OK to me, but I let @zbraniecki review the macro usage.

zbraniecki commented:
Thank you for the PR! The only thing I'd like to ponder over is that you're introducing an example where our sepration of Preferences and Options gets blurry - the developer sets an option, and user may set another. We have to decide who takes precedence.

zbraniecki commented on components/collator/src/options.rs:
```diff
@@ -198,6 +202,27 @@ pub enum CaseFirst {
+            CollationCaseFirst::Upper => CaseFirst::UpperFirst,
+            CollationCaseFirst::Lower => CaseFirst::LowerFirst,
+            CollationCaseFirst::False => CaseFirst::Off,
+            _ => unreachable!("`CaseFirst` and `CollationCaseFirst` must always be in sync."),
```
`q`: Should we panic here or debug_assert and in release select `[default]`?

zbraniecki commented on components/collator/src/options.rs:
```diff
@@ -198,6 +202,27 @@ pub enum CaseFirst {
+            CaseFirst::Off => CollationCaseFirst::False,
+            CaseFirst::LowerFirst => CollationCaseFirst::Lower,
+            CaseFirst::UpperFirst => CollationCaseFirst::Upper,
+        }
```
Why you don't need a catch case despite CaseFirst being non-exhaustive?

Should we switch Options to use CollactionCaseFirst from Preferences then?

jedel1043 replied:
`#[non_exhaustive]` only applies for external crates that use the enum.

I preserved the old enum because I wanted to avoid unnecessary breaking changes, but we can switch to that if preferred.

zbraniecki replied:
We have a shot at doing it in 2.0, I think i'd prefer that. @sffc @Manishearth @robertbastian  - you ok with that?

sffc replied:
I have a slight preference for removing the settings from `options` and moving them to `preferences` in 2.0

Manishearth replied:
Same.

jedel1043 replied:
Just to be clear – we would move the relevant options to `CollatorPreferences`, using the structs on `icu_locale_core`, and the rest of the options would still be in `CollatorOptions`, is that right?

zbraniecki replied:
correct!

jedel1043 commented:
Ok, switched it to use preference keyword structs and removed the old config structs. Had to slightly adjust the documentation to mention `CollatorPreferences`, but it should be feature-complete.

zbraniecki commented:
Oh, interesting. Am I correct to read your PR as - you removed the two preferences from Options, and allow for setting it by the developer by overriding Preferences?

In other words if the developer wants to set one of those two, they interact with Preferences bag, not Options bag anymore?

jedel1043 commented:
@zbraniecki Yes. That way, the problem we had about priority is solved in an arguably cleaner way, since users can either extend the locale preferences to give priority to the explicit preferences, or do the opposite to give priority to the implicit locale preferences.

zbraniecki commented:
> Yes. That way, the problem we had about priority is solved in an arguably cleaner way, since users can either extend the implicit locale preferences with explicit preferences to give priority to the latter, or do the opposite to give priority to the former.

Ok, so you think this is even a cleaner approach! I value your perspective as an implementer. I think I have a hard time deciding on the tradeoffs around it.
My main concern would be that Preferences are intended to be out of control for developers - they accept it (from Locale and from Regional Prefs of their system) and pass it on. The Options on the other hand are meant to be set by the developer. I thought this is a nice clean design but of course reality check is that there are cases where Prefs and Options duplicate and you brought up this here.

With your change we're telling developers to touch Preferences manually. The value of it is that, as you rightly pointed out, it is clean and allows developers to control which direction they want to merge - options->prefs, or prefs->options. ECMA-402 specifies one, but that doesn't mean everyone has to repeat it.

The downsides are:
1) we make the distinction between Prefs and Options murky. "Prefs are for customers, but if you want to set X, modify them, as you do with Options"
2) we're now explicitly requiring implementers to think of this nuance and risk getting it wrong or not at all.

I think I'm fine with that as the benefit outweights the cost for me but wanted to have this explicit paper trail of this decision to revisit in the future.

zbraniecki reviewed (dismissed):
lgtm! thank you!

sffc reviewed (approved):
Need to discuss the re-export in https://github.com/unicode-org/icu4x/issues/5915 but otherwise this looks great; thanks!

zbraniecki commented on components/collator/src/lib.rs:
```diff
@@ -284,12 +321,13 @@ extern crate alloc;
 pub use comparison::Collator;
 pub use comparison::CollatorBorrowed;
 pub use comparison::CollatorPreferences;
+pub use icu_locale_core::preferences::extensions::unicode::keywords::CollationCaseFirst;
+pub use icu_locale_core::preferences::extensions::unicode::keywords::CollationNumericOrdering;
+pub use icu_locale_core::preferences::extensions::unicode::keywords::CollationType;
```
Good catch, I think I also prefer clients importing from icu::locale.

jedel1043 commented:
Ok, made some more changes which include renaming the export from `CollationCaseFirst` to `CaseFirst` on the root of `icu_collator`. This aligns with @sffc's [comment](https://github.com/unicode-org/icu4x/issues/5915#issuecomment-2588597773), and I think it looks a lot cleaner.

sffc commented on components/collator/src/lib.rs:
```diff
@@ -285,12 +322,13 @@ extern crate alloc;
+pub use icu_locale_core::preferences::extensions::unicode::keywords::CollationCaseFirst as CaseFirst;
+pub use icu_locale_core::preferences::extensions::unicode::keywords::CollationNumericOrdering as NumericOrdering;
+pub use icu_locale_core::preferences::extensions::unicode::keywords::CollationType;
 pub use options::AlternateHandling;
```
Observation: in icu_datetime, `options` is a public module/namespace. But icu_collator re-exports the things from the top level.

sffc replied:
https://github.com/unicode-org/icu4x/issues/5991

zbraniecki commented:
Thank you for your work on this @jedel1043 !
